[embedded] Use `__has_feature(swiftcc)` to detect Swift calling convention

kateinoigakukun opened the pull request:
This change replaces the use of `__has_feature(swiftasynccc)` with `__has_feature(swiftcc)` to detect the SwiftCC availability. The former condition works fine for most platforms that support both SwiftCC and SwiftAsyncCC or neither, but it fails for WebAssembly, which supports SwiftCC but not SwiftAsyncCC.

This change depends on https://github.com/llvm/llvm-project/pull/85347

Close https://github.com/apple/swift/issues/72343

kateinoigakukun committed:
[embedded] Use `__has_extension(swiftcc)` to detect Swift calling convention

This change replaces the use of `__has_feature(swiftasynccc)` with
`__has_extension(swiftcc)` to detect the SwiftCC availability. The former
condition works fine for most platforms that support both SwiftCC and
SwiftAsyncCC or neither, but it fails for WebAssembly, which supports
SwiftCC but not SwiftAsyncCC.

We add `swiftcc` extension to Clang, and use it here.

kateinoigakukun committed:
[test] Add a test for class allocation on Embedded Wasm

kubamracek commented:
If https://github.com/llvm/llvm-project/pull/85347/files causes `__has_feature(swiftasynccc)` to stop working, then we also need to adjust Runtime/Config.h.

kateinoigakukun committed:
[runtime] Use `__has_extension` to detect `swiftasynccc` availability

`swiftasynccc` is now an extension instead of a feature, so we should
use `__has_extension` to detect its availability.

kateinoigakukun commented:
@swift-ci Please test with the following PR
https://github.com/apple/llvm-project/pull/8441

kateinoigakukun commented:
@swift-ci Please test WebAssembly
https://github.com/apple/llvm-project/pull/8441

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

kubamracek commented on test/embedded/classes-wasm.swift:
```diff
@@ -0,0 +1,35 @@
+// RUN: %clang -target wasm32-unknown-none-wasm %t/check.o %t/rt.c -nostdlib -o %t/check.wasm
+// RUN: %target-run %t/check.wasm
+// REQUIRES: executable_test
+// REQUIRES: CPU=wasm32
```
Can you point me to a CI job that actually runs these executable WASM tests?

kateinoigakukun replied:
Wasm CI runs executable tests on WasmKit runtime: https://ci.swift.org/job/swift-PR-Linux-crosscompile-wasm/31

aschwaighofer commented on stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h:
```diff
@@ -28,8 +28,7 @@
 
-// FIXME: Replace with __has_feature(swiftcc) once that's added to Clang.
-#if __has_feature(swiftasynccc)
+#if __has_extension(swiftcc)
```
Should we use `__has_feature(swiftasynccc) || __has_extension(swiftcc)` to be compatible with older compilers for a while?

@compnerd Do all windows builds use a just built compiler such that this is not a concern?

kateinoigakukun replied:
I think it's worth being conservative here.

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

kateinoigakukun commented:
Hmm, `#if __has_attribute(swiftasynccall) && (__has_feature(swiftasynccc) || __has_extension(swiftasynccc))` is not acceptable with MSVC cl.exe?

https://ci-external.swift.org/job/swift-PR-windows/23188/console
```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\include\swift/Runtime/Config.h(231): fatal error C1012: unmatched parenthesis: missing ')'
```

```c
  // SWIFT_CC(swiftasync) is the Swift async calling convention.
  // We assume that it supports mandatory tail call elimination.
  #if __has_attribute(swiftasynccall) && (__has_feature(swiftasynccc) || __has_extension(swiftasynccc))
  #define SWIFT_CC_swiftasync __attribute__((swiftasynccall))
  #else
  #define SWIFT_CC_swiftasync SWIFT_CC_swift
  #endif
```

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

kateinoigakukun commented:
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly
